CortenMM

junyang-zh opened the pull request:
Implements #2792

---

Dependencies:
 - #2351
 - #2744
 - #2595
 - #2605
 - #2613
 
 Conflicts with this PR and can be resolved in either way:
 - #2766
 - #2736
 - #2157

Status:

 - [x] fork
 - [x] map
 - [x] page fault
 - [x] protect
 - [x] query
 - [x] remap
 - [x] unmap
 - [x] access_remote
 - [x] ~~Fix mmap races by synchronizing in the allocator~~
 - [x] Fix for tests in #2766
 - [x] Remove the allocator to avoid races
 - [x] make correct the rlimit
 - [x] optimize unfixed map allocation

junyang-zh committed:
Use generics to separate `Cursor` and `CursorMut`

junyang-zh committed:
Add a set of APIs for auxiliary PT metadata

junyang-zh committed:
No automatic level adjusting in `query`/`map`

junyang-zh committed:
Stuff `VmMapping`s into page tables

tatetian commented:
Congratulations for completing all tasks!

I skimmed the code. This PR is huge. If merged as it is now, this PR will be the largest ever. It seems that you have extracted some commits as separate PRs. It would be great if you can update the PR description, listing all PRs that should go before this one.

junyang-zh committed:
Move `VmSpace` tests to another file

junyang-zh committed:
Stop `Cursor` moving after operations

junyang-zh committed:
Replace `Cursor` API's `len` specified range style to `end`

This is to improve the use case in `while cursor.find_next(end)` loops. Ends
are always fixed, while len needs to be calculated in the condition.
